### PR TITLE
Chapter 8

### PR DESCRIPTION
**Pull request by Moonforeva**

### CODE REVIEW

**Comment by Moonforeva** [comment]
Ready

File: app/assets/stylesheets/sessions.scss [hunk 1]
```diff
@@ -0,0 +1,3 @@
+// Place all the styles related to the Sessions controller here.
+// They will automatically be included in application.css.
+// You can use Sass (SCSS) here: https://sass-lang.com/
```
Comment (saobv-2223) [inline app/assets/stylesheets/sessions.scss:3]: remove file that never used

File: app/controllers/sessions_controller.rb [hunk 2]
```diff
@@ -0,0 +1,20 @@
+class SessionsController < ApplicationController
+  def new
+  end
```
Comment (saobv-2223) [inline app/controllers/sessions_controller.rb:3]:
```suggestion
  def new; end
```

File: app/views/layouts/_header.html.erb [hunk 3]
```diff
@@ -5,7 +5,24 @@
       <ul class="nav navbar-nav navbar-right">
         <li><%= link_to "Home", root_path %></li>
         <li><%= link_to "Help", help_path %></li>
-        <li><%= link_to "Log in", "#" %></li>
+        <% if logged_in? %>
+          <li><%= link_to "Users", "#" %></li>
+          <li class="dropdown">
+            <a href="#" class="dropdown-toggle" data-toggle="dropdown">
+              Account <b class="caret"></b>
+            </a>
```
Comment (saobv-2223) [inline app/views/layouts/_header.html.erb:13]: usingg link_to

File: app/views/sessions/new.html.erb [hunk 4]
```diff
@@ -0,0 +1,19 @@
+<% provide :title, "Log in" %>
+<h1>Log in</h1>
+
+<div class="row">
+  <div class="col-md-6 col-md-offset-3">
+    <%= form_for(:session, url: login_path) do |f| %>
```
Comment (saobv-2223) [inline app/views/sessions/new.html.erb:6]: remove ()

File: app/views/sessions/new.html.erb [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+
+      <%= f.label :password %>
+      <%= f.password_field :password, class: 'form-control' %>
+      
+      <%= f.submit "Log in", class: "btn btn-primary" %>
+    <% end %>
+
+    <p>New user? <%= link_to "Sign up now!", signup_path %></p>
+  </div>
+</div>
```
Comment (saobv-2223) [inline app/views/sessions/new.html.erb:19]: add endline

**Comment by Moonforeva** [comment]
Ready

File: app/helpers/sessions_helper.rb [hunk 6]
```diff
@@ -0,0 +1,25 @@
+module SessionsHelper
+
+  #Logs in the given user.
```
Comment (chungpv-1008) [inline app/helpers/sessions_helper.rb:3]: remove this comment

File: app/helpers/sessions_helper.rb [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+  def log_in user
+    session[:user_id] = user.id
+  end
+
+  #Returns the current logged-in user (if any).
```
Comment (chungpv-1008) [inline app/helpers/sessions_helper.rb:8]: similar

File: app/helpers/sessions_helper.rb [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+  def current_user
+    if session[:user_id]
+      @current_user ||= User.find_by id: session[:user_id]
+    end
+  end
+
+  #Returns true if the user is logged in, false otherwise.
```
Comment (chungpv-1008) [inline app/helpers/sessions_helper.rb:15]: similar

File: app/helpers/sessions_helper.rb [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+  def logged_in?
+    !current_user.nil?
```
Comment (chungpv-1008) [inline app/helpers/sessions_helper.rb:17]:
```suggestion
    current_user.present?
```

File: app/helpers/sessions_helper.rb [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+  end
+
+  #logs out the current user.
```
Comment (chungpv-1008) [inline app/helpers/sessions_helper.rb:20]: remove this cmt

**Comment by Moonforeva** [comment]
Ready

File: app/helpers/sessions_helper.rb [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+  end
+
+  def current_user
+    if session[:user_id]
+      @current_user ||= User.find_by id: session[:user_id]
+    end
+  end
+
+  def logged_in?
+    !current_user.present?
```
Comment (manhbnt-0960) [inline app/helpers/sessions_helper.rb:14]:
```suggestion
    current_user.present?
```

File: app/models/user.rb [hunk 12]
```diff
@@ -14,6 +14,13 @@ class User < ApplicationRecord
   
   before_save :downcase_email
   
+  #Returns the hash digest of the given string.
+  def User.digest string
```
Comment (manhbnt-0960) [inline app/models/user.rb:18]: please use syntax `class << self` to define class method

File: config/routes.rb [hunk 13]
```diff
@@ -1,10 +1,14 @@
 Rails.application.routes.draw do
+  get 'sessions/new'
```
Comment (manhbnt-0960) [inline config/routes.rb:2]: remove this line because don't use

File: config/routes.rb [hunk 14]
```diff
@@ -1,10 +1,14 @@
...
   root "static_pages#home"
   
   get   "/home", to: "static_pages#home"       
   get   "/help", to: "static_pages#help"
   get   "/about", to: "static_pages#about"
   get   "/contact", to: "static_pages#contact"
   get   "/signup", to: "users#new"
+  get   "/login", to: "sessions#new"
+  post  "/login", to: "sessions#create"
```
Comment (manhbnt-0960) [inline config/routes.rb:11]:
```suggestion
  post "/login", to: "sessions#create"
```

File: config/routes.rb (hunk 13 above)
Comment (Moonforeva) [reply config/routes.rb:2]: Done

File: config/routes.rb (hunk 14 above)
Comment (Moonforeva) [reply config/routes.rb:11]: Done

File: app/helpers/sessions_helper.rb (hunk 11 above)
Comment (Moonforeva) [reply app/helpers/sessions_helper.rb:14]: Done

**Comment by Moonforeva** [comment]
Ready

File: app/helpers/sessions_helper.rb [hunk 15]
```diff
@@ -0,0 +1,21 @@
+module SessionsHelper
+
```
Comment (saobv-2223) [inline app/helpers/sessions_helper.rb:2]: remove line 2

**Review by saobv-2223 — COMMENTED** [review]
small error syntax, fix it and merge

File: app/helpers/sessions_helper.rb (hunk 15 above)
Comment (Moonforeva) [reply app/helpers/sessions_helper.rb]: Done

**Pull request merged**